pkg/querier/batch/stream.go

```diff
@@ -39,64 +45,120 @@ func (bs *batchStream) atTime() int64 {
 
 func (bs *batchStream) at() (int64, float64) {
 	b := &(*bs)[0]
-	return b.Timestamps[b.Index], b.Values[b.Index]
+	return b.Timestamps[b.Index], b.SampleValues[b.Index]
+}
+
+func (bs *batchStream) atHistogram() (int64, *histogram.Histogram) {
+	b := &(*bs)[0]
+	return b.Timestamps[b.Index], b.HistogramValues[b.Index]
 }
 
+func (bs *batchStream) atFloatHistogram() (int64, *histogram.FloatHistogram) {
+	b := &(*bs)[0]
+	return b.Timestamps[b.Index], b.FloatHistogramValues[b.Index]
+}
+
+// mergeStreams merges streams of Batches of the same series over time.
+// Samples are simply merged by time when they are the same type (float/histogram/...), with the left stream taking precedence if the timestamps are equal.
+// When sample are different type, batches are not merged. In case of equal timestamps, histograms take precedence since they have more information.
 func mergeStreams(left, right batchStream, result batchStream, size int) batchStream {
```

Comment:
Do we have a benchmark here to check if all the function calls indirection introduce a performance penalty? This is an hot path for us.

Reply:
reverted